File: internship/with_java/src/sort/MergeSort.java
package sort;

public class MergeSort {
    // sap xep hop nhat
    static void mergeSort(int arr[], int l, int m, int r) {
        // kich thuoc 2 mang con duoc hop nhat
        int n1 = m - 1 + 1;
        int n2 = r - m;

        // tao mang tam thoi
        int L[] = new int[n1];
        int R[] = new int[n2];

        // copy data to temp arrays
        for (int i = 0; i < n1; ++i)
            L[i] = arr[l + i];
        for (int j = 0; j < n2; ++j)
            R[j] = arr[m + 1 + j];

        // ghep 2 mang tam
        // chi so ban dau cua mang con 1 va 2
        int i = 0, j = 0;

        int k = l;
        while (i < n1 && j < n2) {
            if (L[i] <= R[j]) {
                arr[k] = L[i];
                i++;
            }
            else {
                arr[k] = R[j];
                j++;
            }
            k++;
        }

        while (i < n1) {
            arr[k] = L[i];
            i++;
            k++;
        }

        while (j < n2) {
            arr[k] = R[j];
            j++;
            k++;
        }
    }

    static void sort(int arr[], int l, int r) {
        if (l < r) {
            int m = l + (r - l) / 2;

            sort(arr, l, m);
            sort(arr, m + 1, r);

            mergeSort(arr, l, m, r);
        }
    }
}

File: internship/with_python/sort/mergeSort.py
def merge(arr, left, mid, right):
    n1 = mid - left + 1
    n2 = right -mid

    # Tạo mảng tạm thời L[] và R[]
    L = [0] * n1
    R = [0] * n2
    
    # sao chep du lieu vao 2 mang tam L[] va R[]
    for i in range(n1):
        L[i] = arr[left + i]
    for j in range(n2):
        R[j] = arr[mid + 1 + j]
    i = 0 # chi sang ban dau cua mang con dau tien
    j = 0  # chi sang ban dau cua mang con thu hai
    k = left # chi sang ban dau cua mang con hop nhat
    
    # Hợp nhất lại các mảng tạm thời
    # vào mảng[left..right]
    while i < n1 and j < n2:
        if L[i] <= R[j]:
            arr[k] = L[i]
            i += 1
        else:
            arr[k] = R[j]
            j += 1
        k += 1

    # Sao chép các phần tử còn lại của L[],
    # nếu có
    while i < n1:
        arr[k] = L[i]
        i += 1
        k += 1
    
    # Sao chép các phần tử còn lại của R[],
    # nếu có
    while j < n2:
        arr[k] = R[j]
        j += 1
        k += 1

def merge_sort(arr, left, right):
    if left < right:
        mid = (left + right) // 2

        merge_sort(arr, left, mid)
        merge_sort(arr, mid + 1, right)
        merge(arr, left, mid, right)

def print_list(arr):
    for i in arr:
        print(i, end=" ")
    print()

if __name__ == "__main__":
    arr = [12, 11, 13, 5, 6, 7]
    n = len(arr)
    print("Given array is", end="\n")
    print_list(arr)
    merge_sort(arr, 0, n - 1)
    print("Sorted array is: ", end="\n")
    print_list(arr)
